### PR TITLE
Add npm publish

### PR DESCRIPTION
**Pull request by rickycodes**
**Add npm publish**

<strike>Opening as a draft for now since it relies on: https://github.com/MetaMask/action-npm-publish/pull/1</strike>

This satisfies all team requirements.

This is an example of how `controllers` can consume [`action-npm-publish`](https://github.com/MetaMask/action-npm-publish/) to automatically publish once a release PR is merged.

You can see an example of this working in my fork here: https://github.com/rickycodes/controllers/actions/runs/2017942150

npm publish is failing in my fork simply because my npm token is intentionally incorrect! I _don't_ want to publish from my fork, but we can see it's building a tarball at the correct version etc.

It's a two job workflow where the build is done in one environment and then cached and used to publish in the second so build and publish are isolated.

this also makes use of [github actions environments](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) so we can have requirements for deployment. this makes it so a set of users or a team has to approve the job run (npm publishing in this case).

### CODE REVIEW

File: .github/workflows/publish-npm.yml
Comment (rickycodes) [inline .github/workflows/publish-npm.yml:54]: this will need to be changed to `MetaMask/` before merge

File: .github/workflows/publish-npm.yml
Comment (rickycodes) [inline .github/workflows/publish-npm.yml:56]: this will need to be setup with an NPM automation token. ideally we make this org wide so we can easily consume the same secret in other repos where we want to publsih automatically from.

File: .github/workflows/publish-npm.yml
Comment (rickycodes) [inline .github/workflows/publish-npm.yml:8]: we're already doing this as part of publish release, so we could alternatively just tack this on there (i actually did it that way first), but I think I prefer this.

File: .github/workflows/publish-npm.yml
Comment (Gudahtt) [inline .github/workflows/publish-npm.yml:45]: I guess this is just needed to get `npm` updated? I thought it came pre-installed on these images, but better not to depend upon that if possible.

Comment (rickycodes) [reply .github/workflows/publish-npm.yml:45]: ah! you're right! sorry I think this must have just been pasted from the above job and I neglected to delete

File: .github/workflows/publish-npm.yml
Comment (rekmarks) [reply .github/workflows/publish-npm.yml:8]: If we do it as part of the `publish-release` workflow, we only have to build once. Any particular reason to do it this way?

Comment (rickycodes) [reply .github/workflows/publish-npm.yml:8]: no real reason tbh. I can easily put these two jobs and accomplish the same thing while only building once by moving them over there. I guess that was the original intention?

**Comment by mcmire** [comment]
I can't wait for this to be merged. This will make things so much easier!

File: .github/workflows/publish-npm.yml
Comment (rickycodes) [reply .github/workflows/publish-npm.yml:8]: actually looking at `publish-release` now and it doesn't appear that there's a `yarn build` in there. :thinking:

File: .github/workflows/publish-release.yml
Comment (rickycodes) [inline .github/workflows/publish-release.yml:40]: we'll need to create this `environment` with whatever restrictions we want for publish. we can require reviews from a team of developers and the workflow will not run until approval is met.

File: .github/workflows/publish-release.yml
Comment (rickycodes) [inline .github/workflows/publish-release.yml:50]:
```suggestion
        uses: MetaMask/action-npm-publish@v1.0.0
```

File: .github/workflows/publish-release.yml
Comment (mcmire) [reply .github/workflows/publish-release.yml:40]: That sounds like a good idea. How would we do that?

Comment (mcmire) [reply .github/workflows/publish-release.yml:40]: Ah, just found the link you posted in the description. Would we have to set up a `publish` environment per-repo?

**Review by mcmire — DISMISSED** [review]
I looked over the `publish` action again and this makes sense to me!

File: .github/workflows/publish-release.yml
Comment (rickycodes) [reply .github/workflows/publish-release.yml:40]: > Would we have to set up a publish environment for every repo...

Yes I believe that's the case now, but there's potentially a future where a single environment could be setup at the org level and used in multiple repositories (I'm assuming).

File: .github/workflows/publish-release.yml
Comment (Gudahtt) [inline .github/workflows/publish-release.yml:36]: I was expecting this to just cache the build. It looks like the entire repo, including all node deps, are cached as well. Is that necessary?

Comment (rickycodes) [reply .github/workflows/publish-release.yml:36]: maybe just adding a negate for `node_modules` is good enough here?

**Comment by Gudahtt** [comment]
LGTM! This needs the npm token and the environment to get setup before merging, right?

And for the environment, it would probably be a good idea to ask @danfinlay to setup an "npm publisher" team or sub-team so we don't manually have to select the deployers in each repo.

**Comment by mcmire** [comment]
We're still waiting on the "publisher" team to get set up, right?

**Comment by rickycodes** [comment]
@mcmire we're holding meeting today to get the team/publish environment all set up. once that's in place I'll resolve discussions and merge :)

File: .github/workflows/publish-release.yml
Comment (naugtur) [inline .github/workflows/publish-release.yml:48]: If you want to be super pedantic about it, passing the exact commit hash from one step to another prevents you from publishing stuff in dist along with mismatched package.json coming from a force-push

Comment (Gudahtt) [reply .github/workflows/publish-release.yml:48]: I don't think we need to explicitly pass the commit hash either. It's included in the `github.event.pull_request` event. We can reference it instead of `.ref`.

```suggestion
          ref: ${{ github.event.pull_request.base.sha }}
```

Property name taken from this example in docs: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#webhook-payload-example-32

Comment (Gudahtt) [reply .github/workflows/publish-release.yml:48]: To be honest I am not sure what happens if you use `.ref` and another commit gets pushed before the action is run :thinking: It might only checkout from the repo in the state at the time the event was triggered. But I guess it couldn't hurt to prefer `sha` if we're not sure.

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Review by Gudahtt — APPROVED** [review]

**Pull request merged**
